render-ws-spark-client/src/main/java/org/janelia/render/client/spark/zspacing/ZPositionCorrectionClient.java

```diff
@@ -0,0 +1,206 @@
+            throws IllegalArgumentException {
+        AlignmentPipelineParameters.validateRequiredElementExists("zSpacing",
+                                                                  pipelineParameters.getZSpacing());
+    }
+
+    /** Run the client as part of an alignment pipeline. */
+    @Override
+    public void runPipelineStep(final JavaSparkContext sparkContext,
+                                final AlignmentPipelineParameters pipelineParameters)
+            throws IllegalArgumentException, IOException {
+
+        final Parameters clientParameters = new Parameters();
+        clientParameters.multiProject = pipelineParameters.getMultiProject();
+        clientParameters.zSpacing = pipelineParameters.getZSpacing();
+        deriveAndSolveCrossCorrelationData(sparkContext, clientParameters);
+    }
+
+    private void deriveAndSolveCrossCorrelationData(final JavaSparkContext sparkContext,
+                                                    final Parameters clientParameters)
+            throws IOException {
+
+        final int comparisonRange = clientParameters.zSpacing.getComparisonRange();
+
+        LOG.info("deriveAndSolveCrossCorrelationData: entry, clientParameters={}, comparisonRange={}",
+                 clientParameters, comparisonRange);
+
+        final MultiProjectParameters multiProjectParameters = clientParameters.multiProject;
+
+        final List<StackWithZValues> stackWithAllZValuesList = multiProjectParameters.buildListOfStackWithAllZ();
+
+        final long totalNumberOfZValues = stackWithAllZValuesList.stream()
+                .map(stackWithZ -> (long) stackWithZ.getzValues().size())
+                .reduce(0L, Long::sum);
+
+        final String runName = "run_" + new SimpleDateFormat("yyyyMMdd_HHmmss_SSS").format(new Date()) + "_z_corr";
+
+        deriveCrossCorrelationData(sparkContext,
+                                   clientParameters,
+                                   comparisonRange,
+                                   totalNumberOfZValues,
+                                   multiProjectParameters,
+                                   runName);
+
+        final JavaRDD<StackWithZValues> rddStackWithAllZValues = sparkContext.parallelize(stackWithAllZValuesList);
+
+        final JavaRDD<Integer> rddCompletedSolveCount = rddStackWithAllZValues.map(stackWithZ -> {
+            LogUtilities.setupExecutorLog4j(stackWithZ.toString());
+            final org.janelia.render.client.zspacing.ZPositionCorrectionClient jClient =
+                    clientParameters.buildJavaClient(stackWithZ, comparisonRange, runName, true);
+            final CrossCorrelationData ccData = jClient.loadCrossCorrelationDataSets();
+            jClient.estimateAndSaveZCoordinates(ccData);
+            return 1;
+        });
+
+        final int completedSolveCount = rddCompletedSolveCount.collect().stream().reduce(0, Integer::sum);
+
+        LOG.info("deriveAndSolveCrossCorrelationData: collect RDD");
```

Comment:
I would expect this message before the actual collection step (otherwise I'd like it to read "collected RDD", in which case it is a bit redundant with the next message).

Reply:
This is a bit of a hack.  The Spark driver logging used to write a dynamically updating progress bar that didn't finish with a carriage return.  I got in the habit of adding a superfluous log statement after each collect so that the next  log statement (the one I really wanted to see) would show up on its own new line.  It just checked recent logs and it looks like Spark no longer does this - or at least no longer does it by default.  So, I'll just remove the superfluous log statement.